src/openfermion/measurements/_quantum_subspace_expansion.py

```diff
@@ -0,0 +1,146 @@
+def get_additional_operators(hamiltonian, expansion_operators):
+    """
+    Generate an operator with additional Pauli strings.
+
+    Quantum subspace expansion requires additional Pauli strings.
+    This function generates a QubitOperator with all the Pauli
+    operator to be measured to perform QSE given the expansion
+    operator.
+
+    Args:
+        hamiltonian(QubitOperator): Hamiltonian.
+        stabilizer_list(list): List of QubitOperators.
+
+    Returns:
+        additional_ops(QubitOperator): Operator with all terms needed
+                for QSE with the given expansion operators.
+
+    Notes:
+        Be aware that the coefficient of the additional_ops operator
+        are meaningless.
+        This function is only useful to know which terms are required
+    to be measured to perform QSE.
+
+    Raises:
+        TypeError: if hamiltonian is not QubitOperator.
+        TypeError: if expansion_operators is not an array-like.
+    """
+    if not isinstance(hamiltonian, QubitOperator):
+        raise TypeError('Hamiltonian must be a QubitOperator object.')
+    if not isinstance(expansion_operators, (QubitOperator, list,
+                                            numpy.ndarray)):
+        raise TypeError('List of stabilizers must be an array-like.')
+    if isinstance(expansion_operators, QubitOperator):
+        expansion_operators = list(expansion_operators)
+
+    ham = copy.deepcopy(hamiltonian)
+    additional_ops = QubitOperator()
+
+    for op1 in expansion_operators:
+        for op2 in expansion_operators:
+            additional_ops += op2 * ham * op1
+
+    additional_ops = additional_ops + ham
+    return additional_ops
+
+
+def calculate_qse_spectrum(hamiltonian, expansion_operators,
+                           expectation_values):
+    """
+    Calculate quantum subspace expansion.
+
+    Quantum subspace expansion (QSE) was described in
+    https://arxiv.org/abs/1603.05681.
+    An extension to perform error mitigation based on QSE was proposed in
+    https://arxiv.org/abs/1807.10050.
+    If the expansion operators are symmetries or stabilizers
+    S-QSE is performed.
+
+    This function calculates the quantum subspace expansion of a given
+    Hamiltonian.
+    It is necessary to pass the expectation values of the Hamiltonian and the
+    additional operators.
+
+    Args:
+        hamiltonian(QubitOperator): Hamiltonian a QubitOperator.
+        expansion_operators(list, QubitOperator): List of QubitOperators
+        of the expansion.
+        expectation_values(QubitOperator): Expectation values as QubitOperator.
+
+    Returns:
+        spectrum_qse(list): Energy spectrum after QSE.
+
+    Raises:
+        TypeError: if hamiltonian is not a QubitOperator.
+        TypeError: if expansion_operators is not an array-like or
+                   QubitOperator.
+        TypeError: if expectation_values is not a QubitOperator.
+        ValueError: if length of expanded operator and length of expectation
```

Comment:
This will raise an unwarranted error if the user provides a DM in QubitOperator form with near-zero coefficients and these get trimmed. But, I can see why you would want to have this here, as otherwise the user might not be doing all the measurements they need to do.

I think the key thing to resolve this is to rethink the data structures used here.